seabird_chat_ingest.proto

```diff
@@ -0,0 +1,144 @@
+message JoinChannelChatRequest {
+  string channel_name = 1;
+}
+
+// LeaveChannelChatRequest requests for the bot to leave the channel with the
+// given ID. Either LeaveChannelEvent or FailedChatEvent SHOULD be returned in
+// response if possible.
+message LeaveChannelChatRequest {
+  string channel_id = 1;
+}
+
+// UpdateChannelChatRequest requests for a the channel to be updated with the
+// given metadata. Either ChangeChannelEvent or FailedChatEvent SHOULD be returned
+// in response if possible.
+message UpdateChannelChatRequest {
+  string channel_id = 1;
+  string topic = 2;
+}
+
+// Each ChatRequest has an ID (which can be attached to outgoing events - only
+// the first outgoing event with this ID will be used as the reply) and an inner
+// message type. Having an ID allows us to route an event back to the requestor
+// without having the chat backend implement a service.
+message ChatRequest {
+  string id = 1;
+
+  oneof inner {
+    SendMessageChatRequest send_message = 2;
+    SendPrivateMessageChatRequest send_private_message = 3;
+    JoinChannelChatRequest join_channel = 4;
+    LeaveChannelChatRequest leave_channel = 5;
+    UpdateChannelChatRequest update_channel = 6;
+  }
+}
+
+// HelloChatEvent must be the first event sent by the chat backend. If a
+// different event is sent, the connection will be closed.
+message HelloChatEvent {
+  common.Identity identity = 1;
+
+  common.Backend service_type = 2;
+}
+
+// This is a marker event to respond to seabird-core that performing an action
+// has succeeded. Note that most requests have explicit associated Events, so
+// this should only be used in absence of a matching event.
+message SuccessChatEvent {}
+
+// This is a marker event to respond to seabird-core that performing an action
+// has failed. Note that an action which takes too long will also be marked as
+// failed automatically. The time alotted for an action is not static and may
+// change based on load.
+message FailedChatEvent {
+  string reason = 1;
+}
+
+// JoinChannelEvent will be sent when a user joins a channel. The topic SHOULD
+// be included if possible, but it can be sent in a topic update event later.
+message JoinChannelEvent {
+  string channel_id = 1;
+  string display_name = 2;
+  string topic = 3;
+}
+
+// LeaveChannelEvent will be sent when a user leaves a channel.
+message LeaveChannelEvent {
+  string channel_id = 1;
+}
+
+// ChangeChannelEvent will occur when a channel's display name or topic is
+// updated.
+message ChangeChannelEvent {
+  string channel_id = 1;
+  string display_name = 2;
+  string topic = 3;
+}
+
+// ChatEvent contains all the different event types a chat backend can emit.
+// Note that these are slightly different to the seabird.Event types as channels
+// here will not be mapped to UUIDs and some events are only to support the data
+// that seabird-core tracks.
+//
+// HelloChatEvent is the only message type that is required for a chat backend
+// to work.
```

Comment:
Is there a threshold of failed events that will cause a disconnect?

Reply:
Not currently - I was basing this on IRC where you might not have a confirmation that you did something.

Reply:
Also, specifying you only need to respond to Hello means if we add more event types and backends don't support them they will continue to work.

Reply:
Right, no issues with that. My only concern was calling out whether or not there's a threshold of failed events. I'd include that (even if there isn't one, call it out) in the "failure behaviors" section

Reply:
I'd rather not explicitly define this quite yet - I'd like to revisit this when we start to add more event types because I think a full solution would include the backend specifying which types of events it supports and I don't want to figure that out just yet.

Reply:
Fair, but backend types are orthogonal. At least call out that chat backends can't currently make assumptions around failure events and timeouts.